### PR TITLE
[feature] #1749: no_std support for version, schema and macro libraries

### PR DESCRIPTION
**Pull request by mversic**
### Description of the Change

`data_model` depends on these libraries. Consider this PR a prerequisite for `no_std` support in `data_model`
Manually implemented `Display` trait for `iroha_version::Error` to have it also available in `no_std` context

### Issue

Relates to #1749

### CODE REVIEW

File: schema/Cargo.toml [hunk 1]
```diff
@@ -7,8 +7,8 @@ edition = "2021"
 [dependencies]
 iroha_schema_derive = { version = "=2.0.0-pre.1", path = "derive" }
 
-serde = { version = "1.0", features = ["derive"] }
-fixnum = { version = "0.6", features = ["i64"]}
+serde  = { version = "1.0.132", default-features = false, features = ["derive", "alloc"] }
```
Comment (appetrosyan) [inline schema/Cargo.toml:10]:
```suggestion
serde = { version = "1.0.132", default-features = false, features = ["derive", "alloc"] }
```
This is a very minor æsthetic point, but consistent single space would be very much easier to work with.

File: schema/Cargo.toml [hunk 2]
```diff
@@ -7,8 +7,8 @@ edition = "2021"
...
+fixnum = { version = "0.6.1",   default-features = false, features = ["i64"] }
 
 [dev-dependencies]
-parity-scale-codec = { version = "2", features = ["derive"] }
+parity-scale-codec = { version = "2.3.1", default-features = false, features = ["derive", "full"] }
```
Comment (appetrosyan) [inline schema/Cargo.toml:14]: I wonder if we need `full` here. [This](https://github.com/paritytech/parity-scale-codec/blob/94f5887207af4abfb85e266985f9bd260ee09e04/Cargo.toml#L53) makes me think that it might be better to hide this behind a feature flag and exclude this feature from any code that interacts with `sumeragi`.

File: schema/derive/src/lib.rs [hunk 3]
```diff
@@ -192,7 +192,7 @@ fn metadata_for_enums(data_enum: &DataEnum) -> (Vec<Type>, Expr) {
             );
             quote! {
                 iroha_schema::EnumVariant {
-                    name: stringify!(#name).to_owned(),
+                    name: String::from(stringify!(#name)),
```
Comment (appetrosyan) [inline schema/derive/src/lib.rs:195]: Not a comment, more of a question, is this change necessary for `no-std` support? AFAIK the performance impact is minimal, since the blanket `impl` isn't too bad, but still...

File: version/Cargo.toml [hunk 4]
```diff
@@ -7,19 +7,18 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [features]
-default = ["scale", "json", "derive"]
-scale = ["parity-scale-codec"]
-json = ["serde_json"]
+default = ["std", "derive", "scale", "json"]
+std = ["iroha_macro/std", "parity-scale-codec/std", "thiserror"]
 derive = ["iroha_version_derive"]
+scale = ["parity-scale-codec/full"]
```
Comment (appetrosyan) [inline version/Cargo.toml:13]: This is why I have a question. Ideally, we should either align everything, or nothing. And in most `Cargo.toml` files in this PR the only things that are aligned (and then only within blocks), are the dependencies. Maybe we should roll this back to single space separation?

Comment (appetrosyan) [reply version/Cargo.toml:13]: Or maybe apply alignment consistently to all blocks, not just `dependencies`?

**Review by appetrosyan — COMMENTED** [review]
Overall looks, good, but have a few questions

File: schema/derive/src/lib.rs (hunk 3 above)
Comment (mversic) [reply schema/derive/src/lib.rs:195]: yeah, it's not about the performance, it's about dependencies. With `to_owned` you would also have to include`use alloc::string::ToOwned` when using this macro. So the reason is to reduce the number of required imports. I think you have to import `String` and `Vec` thought

File: version/Cargo.toml (hunk 4 above)
Comment (mversic) [reply version/Cargo.toml:13]: I'll just revert this

File: schema/Cargo.toml (hunk 2 above)
Comment (mversic) [reply schema/Cargo.toml:14]: In Parity Scale this is implied by `std`: `#[cfg(any(feature = "std", feature = "full"))]`

**Review by appetrosyan — DISMISSED** [review]

File: schema/Cargo.toml (hunk 2 above)
Comment (mversic) [reply schema/Cargo.toml:14]: I asked a question with parity, but I find it quite unlikely this should bother us

File: version/src/lib.rs [hunk 5]
```diff
@@ -19,66 +21,81 @@ use serde::{Deserialize, Serialize};
...
+                Self::Serde(_) => "JSON (de)serialization issue",
+                #[cfg(feature = "scale")]
+                Self::ParityScale(_) => "Parity SCALE (de)serialization issue",
+                Self::ParseInt(_) => "Problem with parsing integers",
+                Self::UnsupportedVersion(_) => "Input version unsupported",
+            };
+
+            write!(f, "{}", msg)
+        }
+    }
```
Comment (s8sato) [inline version/src/lib.rs:75]: Maybe we can use [`derive_more`](https://docs.rs/derive_more/latest/derive_more/index.html), which supports `no_std` ?

**Review by s8sato — APPROVED** [review]

File: version/src/lib.rs (hunk 5 above)
Comment (mversic) [reply version/src/lib.rs:75]: Idk what to say. I considered it but decided against it. It doesn't seem to bring much to the table for these small crates.
I can add `derive_more::Display` if you think it would bring something?

Comment (s8sato) [reply version/src/lib.rs:75]: I'd like to accept your decision then

File: schema/Cargo.toml (hunk 2 above)
Comment (mversic) [reply schema/Cargo.toml:14]: Answer from parity:
> It is what I meant with this comment. When building blockchains and using Strings in the runtime shows that you are probably doing something wrong. Like you trying to store user data or whatever. In these cases it is better to hash the data and only store the hash.

**Review by appetrosyan — APPROVED** [review]
